In [1]:
# necessery libraries
import pandas as pd
import sys
sys.path.insert(0,'../src/utils/')
import general_look_at_data as  GLAD

# Explore Dataset in General

In [2]:
# data source must be updated to featsel folder rather than raw
data = pd.read_csv('../data/raw/Citywide_Survey.csv')

In [3]:
# get the shape of the data
data.shape

(8286, 165)

In [4]:
# set option setting to display all columns
pd.set_option('display.max_columns', None)

In [5]:
# Display the head of data
data.head()

hh_id     weight        cms_zone  survey_mode    person_id  person_num  \
0  191546322    0.00000  Inner Brooklyn            1  19154632203           3   
1  191546322    0.00000  Inner Brooklyn            1  19154632202           2   
2  191546322  212.12729  Inner Brooklyn            1  19154632201           1   
3   19241657    0.00000   Middle Queens            1   1924165703           3   
4   19241657    0.00000   Middle Queens            1   1924165702           2   

   is_participant  num_days  num_days_complete  num_trips  num_walk_trips  \
0               0       NaN                NaN        NaN             NaN   
1               0       NaN                NaN        NaN             NaN   
2               1       7.0                7.0       23.0             1.0   
3               0       NaN                NaN        NaN             NaN   
4               0       NaN                NaN        NaN             NaN   

   num_transit_trips  num_bike_trips  num_taxi_trips  num_tnc_trips  \
0                NaN             NaN             NaN            NaN   
1                NaN             NaN             NaN            NaN   
2                3.0            11.0             0.0            0.0   
3                NaN             NaN             NaN            NaN   
4                NaN             NaN             NaN            NaN   

   num_complete_weekend_days  num_complete_weekdays first_travel_date  \
0                        NaN                    NaN        06/24/2019   
1                        NaN                    NaN        06/24/2019   
2                        2.0                    5.0        06/24/2019   
3                        NaN                    NaN        05/29/2019   
4                        NaN                    NaN        05/29/2019   

  last_travel_date  made_trips  smartphone_type  relationship  age  \
0       06/30/2019         NaN              995             5    9   
1       06/30/2019         NaN              995             5    8   
2       06/30/2019         1.0                1             0    5   
3       06/04/2019         NaN              995             3    8   
4       06/04/2019         NaN              995             3    7   

   employment  worker  student  license  vehicle  congestion  drive_cbd_freq  \
0           3       1        0      995      995         995             995   
1           6       0        0      995      995         995             995   
2           6       0        1        1        1           4               7   
3           1       1        0      995      995         995             995   
4           6       0        0      995      995         995             995   

   planning_apps  job_type  jobs_count  industry  telework_freq work_cms_zone  \
0            995       995         995       995            995           NaN   
1            995       995         995       995            995           NaN   
2              1       995         995       995            995           NaN   
3            995       995         995       995            995           NaN   
4            995       995         995       995            995           NaN   

   missing_work_location  work_mode  work_mode_own  work_park_location  \
0                      0        995            995                 995   
1                      0        995            995                 995   
2                      0        995            995                 995   
3                      0        995            995                 995   
4                      0        995            995                 995   

   work_park_pay  work_vehicle_pass_period  work_park_amount_day  \
0            995                       995                   NaN   
1            995                       995                   NaN   
2            995                       995                   NaN   
3            995                       995                   NaN   
4            995 

In [6]:
# check features data types 
data.dtypes.unique()

array([dtype('int64'), dtype('float64'), dtype('O')], dtype=object)

# Explore Categorical vs. Numerical Datasets

In [7]:
# get numeric and categorical features names 
numeric_features = data.select_dtypes(include=['float64', 'int64'])
categorical_features = data.select_dtypes(include='O')

## Statistical Information from Numerical Dataset

In [8]:
# Display 5 random samples of numeric dataset
numeric_features.sample(5)

hh_id    weight  survey_mode   person_id  person_num  is_participant  \
6820  19063533   0.00000            2  1906353302           2               0   
7620  19223247  90.29783            3  1922324701           1               1   
1500  19826025   0.00000            1  1982602502           2               0   
2798  19397431   0.00000            1  1939743103           3               0   
6388  19651319   0.00000            1  1965131903           3               0   

      num_days  num_days_complete  num_trips  num_walk_trips  \
6820       NaN                NaN        NaN             NaN   
7620       1.0                1.0        0.0             0.0   
1500       NaN                NaN        NaN             NaN   
2798       NaN                NaN        NaN             NaN   
6388       NaN                NaN        NaN             NaN   

      num_transit_trips  num_bike_trips  num_taxi_trips  num_tnc_trips  \
6820                NaN             NaN             NaN            NaN   
7620                0.0             0.0             0.0            0.0   
1500                NaN             NaN             NaN            NaN   
2798                NaN             NaN             NaN            NaN   
6388                NaN             NaN             NaN            NaN   

      num_complete_weekend_days  num_complete_weekdays  made_trips  \
6820                        NaN                    NaN         NaN   
7620                        0.0                    1.0         0.0   
1500                        NaN                    NaN         NaN   
2798                        NaN                    NaN         NaN   
6388                        NaN                    NaN         NaN   

      smartphone_type  relationship  age  employment  worker  student  \
6820              995             1    8           1       1        0   
7620                0             0   11           6       0        0   
1500              995             2    3         995     995        1   
2798              995             1    7           1       1        0   
6388              995             2    2         995     995        1   

      license  vehicle  congestion  drive_cbd_freq  planning_apps  job_type  \
6820      995      995         995             995            995       995   
7620        0      995           2             995              8       995   
1500      995      995         995             995            995       995   
2798      995      995         995             995            995       995   
6388      995      995         995             995            995       995   

      jobs_count  industry  telework_freq  missing_work_location  work_mode  \
6820         995       995            995                      0        995   
7620         995       995            995                      0        995   
1500         995       995            995                      0        995   
2798         995       995            995                      0        995   
6388         995       995            995                      0        995   

      work_mode_own  work_park_location  work_park_pay  \
6820            995                 995            995   
7620            995                 995            995   
1500            995                 995            995   
2798            995                 995            995   
6388            995                 995            995   

      work_vehicle_pass_period  work_park_amount_day  work_park_amount_week  \
6820                       995                   NaN                    NaN   
7620                       995                   NaN                    NaN   
1500                       995                   NaN                    NaN   
2798                       995                   NaN                    NaN   
6388                       995                   NaN                    NaN   

      work_park_amount_month  work_park_amount_year  work_

In [9]:
# Display the columns number of numeric dataset
numeric_features.shape[1]

160

In [10]:
# Display Statistics of numeric data
numeric_features.describe()

hh_id        weight  survey_mode     person_id   person_num  \
count  8.286000e+03   8286.000000  8286.000000  8.286000e+03  8286.000000   
mean   2.468398e+07    804.993000     1.246681  2.468398e+09     2.146633   
std    2.974500e+07   4395.258128     0.496917  2.974500e+09     1.317302   
min    1.900024e+07      0.000000     1.000000  1.900024e+09     1.000000   
25%    1.926343e+07      0.000000     1.000000  1.926343e+09     1.000000   
50%    1.951784e+07      0.000000     1.000000  1.951784e+09     2.000000   
75%    1.979496e+07    199.885270     1.000000  1.979496e+09     3.000000   
max    1.999892e+08  88819.119060     3.000000  1.999892e+10    11.000000   

       is_participant     num_days  num_days_complete    num_trips  \
count     8286.000000  3346.000000        3346.000000  3346.000000   
mean         0.403814     5.475792           4.334130    25.540646   
std          0.490691     2.612294           2.524525    21.900434   
min          0.000000     1.000000           1.000000     0.000000   
25%          0.000000     1.000000           1.000000     4.000000   
50%          0.000000     7.000000           5.000000    24.000000   
75%          1.000000     7.000000           7.000000    41.000000   
max          1.000000     7.000000           7.000000   178.000000   

       num_walk_trips  num_transit_trips  num_bike_trips  num_taxi_trips  \
count     3346.000000        3346.000000     3346.000000     3346.000000   
mean        10.004782           5.827555        0.419307        0.111775   
std         12.091553           7.646751        2.141812        1.664201   
min          0.000000           0.000000        0.000000        0.000000   
25%          0.000000           0.000000        0.000000        0.000000   
50%          5.000000           2.000000        0.000000        0.000000   
75%         16.000000          10.000000        0.000000        0.000000   
max         93.000000          57.000000       36.000000       92.000000   

       num_tnc_trips  num_complete_weekend_days  num_complete_weekdays  \
count    3346.000000                3346.000000            3346.000000   
mean        0.374776                   1.266886               3.067244   
std         1.181070                   0.843984               1.858192   
min         0.000000                   0.000000               0.000000   
25%         0.000000                   0.000000               1.000000   
50%         0.000000                   2.000000               4.000000   
75%         0.000000                   2.000000               5.000000   
max        18.000000                   2.000000               5.000000   

        made_trips  smartphone_type  relationship          age   employment  \
count  3346.000000      8286.000000   8286.000000  8286.000000  8286.000000   
mean      0.907651       593.778421   -454.682356     5.718561   179.318972   
std       0.289561       487.541033   2140.005720     2.402113   379.302123   
min       0.000000         0.000000  -9998.000000     1.000000     1.000000   
25%       1.000000         2.000000      0.000000     4.000000     1.000000   
50%       1.000000       995.000000      1.000000     6.000000     3.000000   
75%       1.000000       995.000000      2.000000     8.000000     6.000000   
max       1.000000       995.000000    997.000000    11.000000   995.000000   

            worker      student      license      vehicle   congestion  \
count  8286.000000  8286.000000  8286.000000  8286.000000  8286.000000   
mean    177.439416     0.314868   426.999879   824.355298   427.818489   
std     380.170664     0.529117  1439.775906   374.851421  1439.534122   
min       0.000000     0.000000 -9998.000000     1.000000 -9998.000000   
25%       0.000000     0.000000     1.000000   995.000000     4.000000   
50%       1.000000     0.000000   995.000000   995.000000   995.000000   
75%       1.000000     1.000000   995.000000   995.000000   995.000000   
max     995.000000     2.000000  

### Conclusions Based on the Statistics Presented
* Almost all numerical data is one hot including.
* Missing data values/labels such as -9998 and 995 have an impact on statistics.
* Because work park amount_ and school park amount_ are rare data, their distribution is skewed.

## Statistical Information from Categorical Dataset

In [11]:
# Display 5 samples of categorical dataset
categorical_features.sample(5)

cms_zone first_travel_date last_travel_date   work_cms_zone  \
5821  Outer Brooklyn        06/03/2019       06/09/2019             NaN   
3997  Northern Bronx        06/02/2019       06/08/2019  Manhattan Core   
2005    Inner Queens        06/02/2019       06/08/2019             NaN   
2951  Southern Bronx        06/01/2019       06/07/2019             NaN   
2092  Manhattan Core        06/08/2019       06/14/2019             NaN   

     school_cms_zone  
5821             NaN  
3997             NaN  
2005             NaN  
2951             NaN  
2092             NaN

In [12]:
# Display 5 samples of categorical dataset
categorical_features.shape[1]

5

In [13]:
# Display Statistics of categorical data
categorical_features.describe()

cms_zone first_travel_date last_travel_date   work_cms_zone  \
count             8286              8286             8286            1687   
unique              10                39               40              12   
top     Northern Bronx        06/07/2019       06/13/2019  Manhattan Core   
freq              1032               738              758             815   

       school_cms_zone  
count              123  
unique              10  
top     Manhattan Core  
freq                35

### Conclusions based on the statistics presented
* The 'work_cms_zone' feature contain nearly 50% missing data.
* Because 'scool_cms_zone' feature are rare data, its distribution is skewed.

## Summary of Exploring the two Datasets
- There are 165 numeric features but some of them are a kind of labels for the data such as tnc_mode, pakages, etc. 
- There are 5 categorical feature 2 of them are of type time series. 
- The above mentioned wrong data types will be handled during the feature engineering process. 

# Explore Missing Data Information

In [14]:
feature_missing_stats = GLAD.get_feature_missing(data)
feature_missing_stats

feature  missing_count  missing_percentage  \
0                       hh_id              0                 0.0   
1                      weight              0                 0.0   
2                    cms_zone              0                 0.0   
3                 survey_mode              0                 0.0   
4                   person_id              0                 0.0   
..                        ...            ...                 ...   
160        disability_walking              0                 0.0   
161       disability_mobility              0                 0.0   
162       disability_dressing              0                 0.0   
163  disability_concentrating              0                 0.0   
164           disability_none              0                 0.0   

     dont_know_percentage  no_response_percentage  \
0                     0.0                     0.0   
1                     0.0                     0.0   
2                     0.0                     0.0   
3                     0.0                     0.0   
4                     0.0                     0.0   
..                    ...                     ...   
160                   0.0                     0.0   
161                   0.0                     0.0   
162                   0.0                     0.0   
163                   0.0                     0.0   
164                   0.0                     0.0   

     technological_error_percentage  not_required_percentage  \
0                               0.0                 0.000000   
1                               0.0                 0.000000   
2                               0.0                 0.000000   
3                               0.0                 0.000000   
4                               0.0                 0.000000   
..                              ...                      ...   
160                             0.0                 0.596186   
161                             0.0                 0.596186   
162                             0.0                 0.596186   
163                             0.0                 0.596186   
164                             0.0                 0.596186   

     prefer_not_to_answer_percentage feature_type  
0                           0.000000        int64  
1                           0.000000      float64  
2                           0.000000       object  
3                           0.000000        int64  
4                           0.000000        int64  
..                               ...          ...  
160                         0.024258        int64  
161                         0.024258        int64  
162                         0.024258        int64  
163                         0.024258        int64  
164                         0.024258        int64  

[165 rows x 9 columns]

In [15]:
feature_missing_stats[(feature_missing_stats.drop(columns=['feature', 'feature_type']) > 0).any(1)]

C:\Users\DELL\AppData\Local\Temp\ipykernel_19320\3368692280.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.any and Series.any will be keyword-only.
  feature_missing_stats[(feature_missing_stats.drop(columns=['feature', 'feature_type']) > 0).any(1)]


feature  missing_count  missing_percentage  \
7                    num_days           4940            0.596186   
8           num_days_complete           4940            0.596186   
9                   num_trips           4940            0.596186   
10             num_walk_trips           4940            0.596186   
11          num_transit_trips           4940            0.596186   
..                        ...            ...                 ...   
160        disability_walking              0            0.000000   
161       disability_mobility              0            0.000000   
162       disability_dressing              0            0.000000   
163  disability_concentrating              0            0.000000   
164           disability_none              0            0.000000   

     dont_know_percentage  no_response_percentage  \
7                     0.0                     0.0   
8                     0.0                     0.0   
9                     0.0                     0.0   
10                    0.0                     0.0   
11                    0.0                     0.0   
..                    ...                     ...   
160                   0.0                     0.0   
161                   0.0                     0.0   
162                   0.0                     0.0   
163                   0.0                     0.0   
164                   0.0                     0.0   

     technological_error_percentage  not_required_percentage  \
7                               0.0                 0.000000   
8                               0.0                 0.000000   
9                               0.0                 0.000000   
10                              0.0                 0.000000   
11                              0.0                 0.000000   
..                              ...                      ...   
160                             0.0                 0.596186   
161                             0.0                 0.596186   
162                             0.0                 0.596186   
163                             0.0                 0.596186   
164                             0.0                 0.596186   

     prefer_not_to_answer_percentage feature_type  
7                           0.000000      float64  
8                           0.000000      float64  
9                           0.000000      float64  
10                          0.000000      float64  
11                          0.000000      float64  
..                               ...          ...  
160                         0.024258        int64  
161                         0.024258        int64  
162                         0.024258        int64  
163                         0.024258        int64  
164                         0.024258        int64  

[152 rows x 9 columns]

In [16]:
feature_missing_stats[feature_missing_stats.missing_percentage>0].iloc[: , :3]

feature  missing_count  missing_percentage
7                    num_days           4940            0.596186
8           num_days_complete           4940            0.596186
9                   num_trips           4940            0.596186
10             num_walk_trips           4940            0.596186
11          num_transit_trips           4940            0.596186
12             num_bike_trips           4940            0.596186
13             num_taxi_trips           4940            0.596186
14              num_tnc_trips           4940            0.596186
15  num_complete_weekend_days           4940            0.596186
16      num_complete_weekdays           4940            0.596186
19                 made_trips           4940            0.596186
35              work_cms_zone           6599            0.796404
42       work_park_amount_day           8262            0.997104
43      work_park_amount_week           8274            0.998552
44     work_park_amount_month           8264            0.997345
45      work_park_amount_year           8282            0.999517
56            school_cms_zone           8163            0.985156
63     school_park_amount_day           8279            0.999155
64   school_park_amount_month           8278            0.999035
65    school_park_amount_week           8281            0.999397
66    school_park_amount_year           8278            0.999035

## Suggestion to Handle Missing Data
1. Based on the previous statistics, I recommend that the **following features be deleted that have a missing percentage of higher than 0.75**:
['work_park_amount_day', 'work_park_amount_week', 'work_park_amount_month', 'work_park_amount_year', 'school_cms_zone', 'school_park_amount_day', 'school_park_amount_month', 'school_park_amount_week', 'school_park_amount_year'].

2. I recommend that the missing data for the **rest of features be filled by the median column value** because it is numeric type data and to not follow noisy data.